litgpt/finetune/adapter.py

```diff
@@ -294,13 +291,20 @@ def fit(
                 save_hyperparameters(setup, checkpoint_file.parent)
                 save_prompt_style(data.prompt_style, checkpoint_file.parent)
 
+        if step_count == max_steps:
+            fabric.print("Final validation ...")
+            train_loss = validate(fabric, model, train_dataloader, tokenizer, eval, data, print_output=False)
+            val_loss = validate(fabric, model, val_dataloader, tokenizer, eval, data, print_output=False)
+            fabric.print(f"Final train loss: {train_loss:.3f} | final val loss: {val_loss:.3f}")
+
 
 # the adapter "kv cache" cannot be initialized under `inference_mode`
 @torch.no_grad()
 def validate(
-    fabric: L.Fabric, model: GPT, val_dataloader: DataLoader, tokenizer: Tokenizer, eval: EvalArgs, data: DataModule
+    fabric: L.Fabric, model: GPT, val_dataloader: DataLoader, tokenizer: Tokenizer, eval: EvalArgs, data: DataModule, print_output: bool = True
 ) -> torch.Tensor:
-    fabric.print("Validating ...")
+    if print_output:
+        fabric.print("Validating ...")
```

Comment:
One SWE tip is to avoid this pattern. Instead add the print to the callers that have `print_output=True` and remove the argument.

Similarly, the instruction part could become a separate function. That is if we want to keep it because we have a lot of users confused by it

Reply:
The problem here is that these are two different validation prints. The first one is to signal that the model enters the validation (this may take a while if the validation set is very large). The second one is checking the prompt. If you move the `fabric.print("Validating ...")` then the print would occur when the validation has already happened.

```bash
@torch.no_grad()
def validate(
    fabric: L.Fabric, model: GPT, val_dataloader: DataLoader, tokenizer: Tokenizer, eval: EvalArgs, data: DataModule, print_output: bool = True
) -> torch.Tensor:
    if print_output:
        fabric.print("Validating ...")
    model.eval()
    losses = torch.zeros(min(len(val_dataloader), eval.max_iters))
    for k, batch in enumerate(val_dataloader):
        if k >= eval.max_iters:
            break
        input_ids, targets = batch["input_ids"], batch["labels"]
        logits = model(input_ids)
        losses[k] = chunked_cross_entropy(logits[..., :-1, :], targets[..., 1:], chunk_size=0)

    val_loss = losses.mean()

    if print_output:
        # produce an example:
        instruction = "Recommend a movie for me to watch during the weekend and explain the reason."
        fabric.print(instruction)
        prompt = data.prompt_style.apply(instruction)
        encoded = tokenizer.encode(prompt, device=fabric.device)
        with fabric.init_tensor():
            # do not set `max_seq_length=max_returned_token` because memory is not a concern here
            model.set_kv_cache(batch_size=1)
        output = generate(
            model, encoded, max_returned_tokens=len(encoded) + eval.max_new_tokens, temperature=0.8, eos_id=tokenizer.eos_id
        )
        model.clear_kv_cache()
        output = tokenizer.decode(output)
        fabric.print(output)

    model.train()
    return val_loss
```